spock-core/src/main/java/org/spockframework/compiler/DeepBlockRewriter.java

```diff
@@ -145,11 +148,39 @@ private boolean handleImplicitCondition(ExpressionStatement stat) {
     checkIsValidImplicitCondition(stat);
     conditionFound = true;
     groupConditionFound = currSpecialMethodCall.isGroupConditionBlock();
+
+    if (currSpecialMethodCall.isWithCall()
+      && AstUtil.isInvocationWithImplicitThis(stat.getExpression())) {
+      replaceObjectExpressionWithCurrentClosure(stat);
+    }
+
     Statement condition = ConditionRewriter.rewriteImplicitCondition(stat, resources);
     replaceVisitedStatementWith(condition);
     return true;
   }
 
+  private void replaceObjectExpressionWithCurrentClosure(ExpressionStatement stat) {
+    MethodCallExpression methodCall = AstUtil.getExpression(stat, MethodCallExpression.class);
+    if (methodCall == null) return;
+
+    MethodCallExpression target = referenceToCurrentClosure();
+    methodCall.setObjectExpression(target);
+  }
+
+  @NotNull
```

Comment:
I see that "org.jetbrains:annotations:13.0" is an optional dependency, but I don't see any other usage of those annotations in code. @robfletcher @sbglasius What do you think about that?